src/stats.py

```diff
@@ -76,40 +48,65 @@ def save_pkl_files_from_kafka(kafka_server, run_id):
+    parser.add_argument("--files", nargs="+",
+                        type=str, help="List of stat pickle files")
     parser.add_argument("--run-id", type=str, required=True)
     parser.add_argument("--result-dir", type=str, required=True)
     opt = parser.parse_args()
 
     # Create result directory
     create_dir(opt.result_dir)
 
-    unpkl_msgs = save_pkl_files_from_kafka(
-        opt.kafka_server, opt.run_id
-    )
-
-    models = [None] * len(unpkl_msgs)
-    metrics = {}
-    for metric_name in METRIC_NAMES:
-        metrics[metric_name] = [None] * len(unpkl_msgs)
-
-    for (tid, infer_stats) in unpkl_msgs:
-        model, tput, total_times, queued_times = infer_stats
-        populate_stats(TOTAL_PREFIX, total_times, tid, metrics)
-        populate_stats(QUEUED_PREFIX, queued_times, tid, metrics)
-        processing_times = [x - y for x, y in zip(total_times, queued_times)]
-        populate_stats(PROCESSING_PREFIX, processing_times, tid, metrics)
-        metrics[TPUT][tid] = tput
-        if models[tid] is None:
-            models[tid] = f"{tid}_{model}"
-        else:
-            assert models[tid] == f"{tid}_{model}"
-
-    # Create a DataFrame for each metric type
-    models = [x for x in models if x is not None]
+    # Get stats across a particular run-id
+    collected_stats = collect_stat_objects(opt.files)
+
+    # Aggreate across spawns of (tid, model) pair
+    models = {}
+    tids_to_spawns = defaultdict(set)
+    for stat_obj in collected_stats:
+        # Make sure they all are for the same run id
+        assert stat_obj.run_id == opt.run_id
+        tids_to_spawns[stat_obj.tid].add(stat_obj.spawn_id)
+        if stat_obj.key not in models:
+            models[stat_obj.key] = Stats(
+                stat_obj.run_id, stat_obj.tid, stat_obj.model_id
+            )
+        models[stat_obj.key].aggregate(stat_obj)
+
+    # Make sure all tids have dumped
+    for i in range(len(tids_to_spawns)):
+        assert i in tids_to_spawns
+
+    # Make sure all spawns have dumped
+    expected_spawns = len(tids_to_spawns[0])
+    for _, spawns in tids_to_spawns.items():
+        assert len(spawns) == expected_spawns
+        for i in range(len(spawns)):
+            assert i in spawns
+
+    model_names = [None] * len(models)
+    metrics = defaultdict(lambda: [None] * len(models))
+
+    # Create ordered model names and metrics
+    for key, stat_obj in models.items():
+        model_names[stat_obj.tid] = key
+
+        # Get throughput metrics
+        metrics["tput"][stat_obj.tid] = stat_obj.tput()
+
+        # Get percentile metrics
+        for metric_type in PERCENTILE_METRIC_TYPES:
+            percentile_metrics = stat_obj.percentile_stats(
+                metric_type, PERCENTILES
+            )
+            for i, p in enumerate(PERCENTILES):
+                metric_key = f"{metric_type}_p{p}"
+                metrics[metric_key][stat_obj.tid] = percentile_metrics[i]
+
```

Comment:
How is the directory structure being ordered? Will I need to always submit jobs of a job mix in a certain order?